### ISSUE TITLE
DNBSEQ-T7 platform demultiplexing

### ISSUE THREAD

**Issue by Maple-sakura** [MGI-tech-bioinformatics/DNBelab_C_Series_HT_scRNA-analysis-software#90]
Hello.
I have 4 snATAC-seq samples from the DNBSEQ-T7 platform, and for each sample, they are mixed with multiple cell lines. For R1, the read length is 50bp, and for R2, the read length is 70bp.
I used PISA to add 51-60 (10bp), 61-70 (10bp) of R2, a total of 20bp, as barcodes to R1 and R2, and extracted the original R1 (50bp), R2 (1-50bp in 70bp) as paired fastq and further used bwa for alignment.
Strangely, the alignment rates of the 4 samples are only close to 50% (46, 48, 53, 54%). I would like to ask if I have misunderstood the position of the barcode?

I also used dnbc4atac for alignment, the command is as follows
dnbc4atac run --fastq1 /my_disk/public/Data/scATAC_RNA/CNS0798658/snATAC_pool4_read_1.fq.gz \
    --fastq2 /my_disk/public/Data/scATAC_RNA/CNS0798658/snATAC_pool4_read_2.fq.gz \
    --name CNS0798658 \
    --genomeDir /home/jiangr/reference/GRch38/dnbc4tools \
    --outdir /mnt/data1/jiangr/CL_scATAC/CNS0798658/dnbc4_result \
    --threads 20

Error log in 01.data/alignment_report.tsv
Less than 5% barcodes can be found or corrected based on the barcode whitelist.

It seems that dnbc4atac cannot handle such data. 
Thank you for your help.

**Comment by lishuangshuang0616** [comment]
The sequencing length you have does not align with MGI's single-cell ATAC reagents. 
MGI's cell barcode should reside within the first 20bp of R1, with a sequencing length of **R1 70bp + R2 50bp**. 
You can initially BLAST a few paired-end reads to verify the barcode's exact position based on alignment information.
 If the barcode is indeed located in the last 20bp of R2, use 
```
dnbc4tools atac run 
   --fastq1 snATAC_pool4_read_2.fq.gz 
   --fastq2 snATAC_pool4_read_1.fq.gz 
   --customize bc:50:69,r1:0:49,r2:0:49
```
 If it's in the first 20bp of R1, you might need to re-split and trim your data to ensure a sequencing length of **R1 70bp + R2 50bp**.

**Comment by Maple-sakura** [comment]
Thanks for your suggestions. I made more attempts to extract the barcode, details are as follows.

### check structure of R1(70bp? read_2)/R2(50bp? read_1)?
`zcat /my_disk/public/Data/scATAC_RNA/CNS0798659/snATAC_pool5_read_2.fq.gz|head -n 8`                                                     
@DP8400011418BRL1C001R0030000004/2
ATATTAGTTCATATTGTTTTTATAGTAATGTTAATAGTGATTAACGCCGTACCATACAATAAGTTGGTGG
+
D;CEE'6@<*AD;A;=)CF'3<.?FB>6?*?@C=FEFCEC?DEAEFE)FCE7FBFGDFGFEFEEDC?'C4
@DP8400011418BRL1C001R0030000016/2
TACTAGGCCTACTAACCAACACACTAACCATATACCAATGATGGCGCGATGAGACTATAGTCACGAATCC
+
IIHGHIHHHIIGGHGHIIIGIIIIIIHHHHHHHIHIHGGI:HIHIIIHGIH<IGH?I4IHHIIIIGIIIG

`zcat /my_disk/public/Data/scATAC_RNA/CNS0798659/snATAC_pool5_read_1.fq.gz|head -n 8`
@DP8400011418BRL1C001R0030000004/1
CGTTAATCACTATTAAAATTACTATAAAAACAATATGAACTAATATCTGG
+
CE>BDCECEDCEEDDEDD=DFADEEFDDDEBEEAE:GCD@BGE>D@E?E&
@DP8400011418BRL1C001R0030000016/1
GGTGGTGTGTGGTGGCCTTGGTATGTGCTTTCTCGTGTTACATCGCGCCA
+
FGCGEIIFHFIGGGAFGIFGFHIIIFHHHFIIGHHEIFGIGEIIHHGIHI

**The fq file name provided in the article does not seem to be correct, since R1=50bp and R2=70bp? However, the fq header is consistent with the fq file name.**

### Use PISA to extract barcode but fail.
`PISA parse2  -t 10 -rule 'CR,R1:1-10,/home/jiangr/tools/PISA-1.0b/config/DNBelabC4_barcodes.txt,CB,1;CR,R1:11-20,/home/jiangr/tools/PISA-1.0b/config/DNBelabC4_barcodes.txt,CB,1;R1,R2:1-100' -1 /mnt/data2/jiangr/CL_scATAC3/CNS0798659/clean/CNS0798659_demultiplex.fq /my_disk/public/Data/scATAC_RNA/CNS0798659/snATAC_pool5_read_2.fq.gz /my_disk/public/Data/scATAC_RNA/CNS0798659/snATAC_pool5_read_1.fq.gz`

DNBelabC4_barcodes.txt contains 10bp barcode list (n=1536). However, the size of CNS0798659_demultiplex.fq is only **177MB**.
/my_disk/public/Data/scATAC_RNA/CNS0798659/snATAC_pool5_read_2.fq.gz (40G); /my_disk/public/Data/scATAC_RNA/CNS0798659/snATAC_pool5_read_1.fq.gz (27G).


### re-run dnbc4atac but fail 

```
dnbc4atac run --fastq1 /my_disk/public/Data/scATAC_RNA/CNS0798658/snATAC_pool4_read_2.fq.gz \
        --fastq2  /my_disk/public/Data/scATAC_RNA/CNS0798658/snATAC_pool4_read_1.fq.gz \
        --name CNS0798658 \
        --genomeDir /home/jiangr/reference/GRch38/dnbc4tools \
        --outdir /mnt/data1/jiangr/CL_scATAC/CNS0798658/dnbc4_result \
        --threads 20
```
When running dnbc4atac, I changed the R1/R2 files. Originally R1(read_1=50bp), R2(read_2=70bp), now R1(read_2=70bp), R2(read_1=50bp), but still not enough barcodes are detected?

Error log in 01.data/alignment_report.tsv
**Less than 5% barcodes can be found or corrected based on the barcode whitelist.**

**I think this R1=50; R2=70bp library is very strange.**

### In summary, I tested two different processing methods.

**Case 1**, 1-20 of 70bp fastq as barcode; PISA only retained a few reads in R1/R2 (fq=177MB).

**Case 2**, 51-70 of 70bp fastq as barcode; CNS0798658_demultiplex.trimed.R1.fq.gz (**originally 40G, now 14G**)
CNS0798658_demultiplex.trimed.R2.fq.gz (**originally 27G, now 14G**), but the mapping rate of the bwa tool is only 50%.

Could you provide some suggestions or code? Thanks.

**Comment by lishuangshuang0616** [comment]
```
PISA parse2  -t 10 \
  -rule 'CR,R2:51-60,/home/jiangr/tools/PISA-1.0b/config/DNBelabC4_barcodes.txt,CB,1;CR,R2:61-70,/home/jiangr/tools/PISA-1.0b/config/DNBelabC4_barcodes.txt,CB,1;R1,R1:1-50;R2,R2:1:50'  \
  -report /mnt/data2/jiangr/CL_scATAC3/CNS0798659/clean/fastq.csv \
  -1 /mnt/data2/jiangr/CL_scATAC3/CNS0798659/clean/snATAC_pool5_read_1.fq  \
  -2 /mnt/data2/jiangr/CL_scATAC3/CNS0798659/clean/snATAC_pool5_read_2.fq \
  /my_disk/public/Data/scATAC_RNA/CNS0798659/snATAC_pool5_read_1.fq.gz  \
  /my_disk/public/Data/scATAC_RNA/CNS0798659/snATAC_pool5_read_2.fq.gz
```

I suggest you try to analyze it in this way. I remember you posted an issue about the article data before. 
I said at that time that the data was a very early version. You see if this analysis is feasible.

**Comment by lishuangshuang0616** [comment]
dnbc4tools need add ```--customize bc:50:69,r1:0:49,r2:0:49```
